### PR TITLE
[PATCH v5] test: sched_pktio: cleanup prefetched events before worker exit

### PR DESCRIPTION
**Pull request by nithind1988**
Cleanup prefetched events before worker exits. Also break before
schedule multi is called if exit is triggered.

Signed-off-by: Nithin Dabilpuram <ndabilpuram@marvell.com>

### CODE REVIEW

File: test/performance/odp_sched_pktio.c [hunk 1]
```diff
@@ -238,6 +238,26 @@ static int worker_thread_direct(void *arg)
...
+		if (ev == ODP_EVENT_INVALID)
+			break;
+		odp_event_free(ev);
+	}
+
+	/* Flush all writes */
+	odp_mb_full();
+
+	/* Restore scheduling state */
+	odp_schedule_resume();
```
Comment (JannePeltonen) [inline test/performance/odp_sched_pktio.c:259]: So after this odp_schedule_resume() call there can again be local prefetched events and we are back to square one? I do not see how this patch fixes something, other than by chance through altering timing.

Maybe there should be thread barrier (like in l2fwd) to make sure everybody stops sending packets before anybody resumes scheduling again for the final flush (if it is needed anymore then). Or something like that.

Comment (nithind1988) [reply test/performance/odp_sched_pktio.c]: Removed odp_schedule_resume() in V2. Let me know if it makes sense given that main thread is cleaning up all the other events in scheduler so I believe odp_l2fwd like per thread cleanup is not needed.

File: test/performance/odp_sched_pktio.c [hunk 2]
```diff
@@ -238,6 +238,24 @@ static int worker_thread_direct(void *arg)
 			test_global->worker_stat[worker_id].rx_pkt += num_pkt;
 	}
 
+	/*
+	 * Free prefetched packets before exiting worker thread as
+	 * such packets can block main thread cleanup cause buffer leak.
```
Comment (JannePeltonen) [inline test/performance/odp_sched_pktio.c:243]: There is a grammar issue in the sentence, making it hard to understand.

File: test/performance/odp_sched_pktio.c [hunk 3]
```diff
@@ -238,6 +238,24 @@ static int worker_thread_direct(void *arg)
...
+		odp_event_t  ev;
+
+		ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+		if (ev == ODP_EVENT_INVALID)
+			break;
+		odp_event_free(ev);
+	}
+
+	/* Flush all writes */
+	odp_mb_full();
```
Comment (JannePeltonen) [inline test/performance/odp_sched_pktio.c:256]: I wonder if and why this is needed. I would have thought thread exit followed by odph_odpthreads_join() ensures the needed memory synchronization between the threads.

File: test/performance/odp_sched_pktio.c (hunk 1 above)
Comment (JannePeltonen) [reply test/performance/odp_sched_pktio.c]: After the removal of the schedule_resume() this looks ok to me, apart from the minor unrelated comments I made.

**Comment by JannePeltonen** [comment]
The commit message could say that this change is only for the case when pipelining or timeouts are not used. I suppose the pipeline and timeout variants of the worker function have the same problems.

File: test/performance/odp_sched_pktio.c (hunk 2 above)
Comment (nithind1988) [reply test/performance/odp_sched_pktio.c:243]: Ack, will fix in V2.

File: test/performance/odp_sched_pktio.c (hunk 3 above)
Comment (nithind1988) [reply test/performance/odp_sched_pktio.c:256]: Ack, Will remove this in V2. Join should be doing it.

**Review by JannePeltonen — APPROVED** [review]
Reviewed-by: Janne Peltonen <janne.peltonen@nokia.com>

**Comment by matiaselo** [comment]
@nithind1988 Please add Janne's reviewed-by tag, so we can merge this PR.

**Pull request merged**